malsim/sims/mal_sim_agent_state.py

```diff
@@ -0,0 +1,87 @@
+"""MalSimAgentState stores the state of an agent inside the MalSimulator"""
+
+from __future__ import annotations
+from typing import TYPE_CHECKING
+from enum import Enum
+from dataclasses import dataclass, field
+
+if TYPE_CHECKING:
+    from maltoolbox.attackgraph import AttackGraphNode
+
+
+class AgentType(Enum):
+    """Enum for agent types"""
+    ATTACKER = 'attacker'
+    DEFENDER = 'defender'
+
+
+@dataclass
+class MalSimAgentState:
+    """Stores the state of an agent in the simulator"""
+
+    # Identifier of the agent, used in MalSimulator for lookup
+    name: str
+    type: AgentType
+
+    # Contains current agent reward in the simulation
+    # Attackers get positive rewards, defenders negative
+    reward: int = 0
+
+    # Contains possible actions for the agent in the next step
+    action_surface: list[AttackGraphNode] = field(default_factory=list)
+
+    # Fields that tell if agent is 'dead' / disabled
+    truncated: bool = False
+    terminated: bool = False
+
+    # Fields that are not used by 'base' MalSimulator
+    # but can be useful in wrappers/envs
+    observation: dict = field(default_factory=dict)
+    info: dict = field(default_factory=dict)
+
+class MalSimAttackerState(MalSimAgentState):
+    """Stores the state of an attacker in the simulator"""
+    def __init__(self, name: str, attacker_id: int):
+        super().__init__(name, AgentType.ATTACKER)
+        self.attacker_id = attacker_id
+
+class MalSimDefenderState(MalSimAgentState):
+    """Stores the state of a defender in the simulator"""
+    def __init__(self, name: str):
+        super().__init__(name, AgentType.DEFENDER)
+
+
+class MalSimAgentStateView:
+    """Read-only interface to MalSimAgentState."""
+    def __init__(self, agent: MalSimAgentState):
+        self._agent = agent
+
+    @property
+    def name(self) -> str:
+        return self._agent.name
+
+    @property
+    def type(self) -> AgentType:
+        return self._agent.type
+
+    @property
+    def reward(self) -> int:
+        return self._agent.reward
+
+    @property
+    def truncated(self) -> bool:
+        return self._agent.truncated
+
+    @property
+    def terminated(self) -> bool:
+        return self._agent.terminated
+
+    @property
+    def action_surface(self) -> list[AttackGraphNode]:
+        return self._agent.action_surface
+
+    @property
+    def attacker_id(self) -> int:
+        assert isinstance(self._agent, MalSimAttackerState), \
+               "Only MalSimAttackers have property attacker_id"
+        return self._agent.attacker_id
```

Comment:
A simpler implementation would be:

```suggestion
@dataclass
class MalSimAgentStateView(MalSimAgentState):
    """Read-only interface to MalSimAgentState."""
    def __post_init__(self) -> None:
        self.__setattr__ = self._freeze
        
    def _freeze(self, name, value) -> None:
        raise AttributeError("State view properties are read-only")
```

I haven't run this so it may need a few tweaks.

Reply:
You can skip defining `_freeze` and use a lambda to save 2 lines.

Reply:
Sth to note is that while the properties of the agent state are read-only, the values of the properties - if mutable - are not. For example, the action surface is not read only itself, the agent can modify it if they want. I think it would be a good idea to lock these down too, i.e. convert the action surface into a tuple of nodes instead of a list.